nomt/src/cursor.rs

```diff
@@ -229,8 +279,121 @@ impl PageCacheCursor {
+                ref mut write_pass,
+                ref mut dirtied,
+            } => (write_pass, dirtied),
+        };
+        let fetched_page;
+
+        let (page_id, page, (left_idx, right_idx)) = match self.cached_page {
+            None => {
+                fetched_page = self.pages.retrieve_sync(ROOT_PAGE_ID);
+                (ROOT_PAGE_ID, &fetched_page, (0, 1))
+            }
+            Some((page_id, ref page)) => {
+                let path = last_page_path(&self.path, self.depth);
+                match child_node_indices(path) {
+                    Some((left, right)) => (page_id, page, (left, right)),
+                    None => {
+                        let child_page_idx = path.load_be::<u8>();
+
+                        // unwrap: this is true even for the leaf node placement, as less than the
+                        // entire 42nd page is used with 256 bit keys.
+                        let child_page_id = page_id.child_page_id(child_page_idx).expect(
+                            "Child index is 6 bits and Pages do not go deeper than the maximum layer, 42"
+                        );
+                        fetched_page = self.pages.retrieve_sync(child_page_id);
+                        (child_page_id, &fetched_page, (0, 1))
+                    }
+                }
+            }
+        };
+
+        page.set_node(&mut *write_pass.borrow_mut(), left_idx, leaf_data.key_path);
+        page.set_node(
+            &mut *write_pass.borrow_mut(),
+            right_idx,
+            leaf_data.value_hash,
+        );
+        dirtied.insert(page_id);
+    }
+
+    /// Place a non-leaf node at the current location.
+    pub fn place_non_leaf(&mut self, node: Node) {
+        if trie::is_leaf(&node) {
+            return;
+        }
+
+        // hack: this assumes that leaves are always explicitly deleted
+        // before being overwritten in order to maintain a consistent state.
+        if trie::is_terminator(&node) && trie::is_leaf(&self.node()) {
+            self.write_leaf_children(Default::default());
+        }
+
+        let (write_pass, dirtied) = match self.mode {
+            Mode::Read(_) => panic!("attempted to call modify on a read-only cursor"),
+            Mode::Write {
+                ref mut write_pass,
+                ref mut dirtied,
+            } => (write_pass, dirtied),
+        };
+        match self.cached_page {
+            None => {
+                self.root = node;
+            }
+            Some((page_id, ref mut page)) => {
+                let path = last_page_path(&self.path, self.depth);
+                let index = node_index(path);
+                page.set_node(&mut *write_pass.borrow_mut(), index, node);
+                dirtied.insert(page_id);
+            }
+        }
+    }
+
+    /// Place a leaf node at the current location.
+    pub fn place_leaf(&mut self, node: Node, leaf: trie::LeafData) {
+        if !trie::is_leaf(&node) {
+            return;
```

Comment:
why not panic? shouldn't we try to make the system as tight as possible, and therefore take the every opportunity to make the fuzz or test fail?